syntax = "proto3";

import "src/api/artifacts/uuid.proto";
import "src/api/artifacts/math.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/any.proto";
import "google/protobuf/field_mask.proto";

package observation_tools.proto;

option java_multiple_files = true;
option java_package = "tools.observation.proto";

message ArtifactId {
  Uuid uuid = 1;
}

message ArtifactData {
  // Ordered from root -> leaf.
  repeated ArtifactId ancestor_group_ids = 6;

  ArtifactUserMetadata user_metadata = 7;

  ArtifactType artifact_type = 8;

  oneof type_data {
    RunStageData run_stage_data = 9;
    Map2dTo3dData map_2d_to_3d = 10;
    Group3d group_3d = 11;
    SeriesData series = 12;
  }

  .google.protobuf.Timestamp client_creation_time = 5;

  reserved 1 to 4;
}

message ArtifactUserMetadata {
  string name = 1;

  map<string, string> metadata = 2;
}

enum ArtifactType {
  ARTIFACT_TYPE_UNKNOWN = 0;
  ARTIFACT_TYPE_GENERIC = 1;
  // Group of 2d objets
  ARTIFACT_TYPE_2D_GROUP = 2;
  // Group of 3d objects
  ARTIFACT_TYPE_3D_GROUP = 3;
  // Transforms a 2D objects into 3d space
  ARTIFACT_TYPE_2D_IN_3D_GROUP = 4;
  ARTIFACT_TYPE_RUN_STAGE = 5;
  ARTIFACT_TYPE_ROOT_GROUP = 6;
  ARTIFACT_TYPE_SERIES = 7;
}

message RunStageData {
  repeated ArtifactId previous_run_stage_ids = 1;
}

message Map2dTo3dData {
  Transform3 to_3d_transform = 2;

  reserved 1;
}

message Group3d {
  Transform3 base_transform = 2;

  reserved 1;
}

message StructuredData {
  oneof data {
    Polygon2 polygon2 = 1 [deprecated = true];
    Graph graph = 2;
    Point2 point2 = 3 [deprecated = true];
    Image2 image2 = 4;
    Segment2 segment2 = 5 [deprecated = true];
    Object3 object3 = 6;
    Object2 object2 = 7;
    Sphere sphere = 8 [deprecated = true];
  }
}

message Image2 {
  bytes data = 4;

  string mime_type = 5;

  // Maybe deprecated
  Matrix3x3 pixel_to_world_transform = 2 [deprecated = true];

  PerPixelTransform pixel_transform = 3;

  reserved 1;
}

message PerPixelTransform {
  RandomDistinctColor random_distinct_color = 1;
}

message RandomDistinctColor {

}

message Object2 {
  Geometry2 geometry = 1;

  repeated Transform2 transforms = 2;
}

message Geometry2 {
  oneof data {
    Point2 point2 = 1;
    Polygon2 polygon2 = 2;
    Segment2 segment2 = 3;
    Image2 image2 = 4;
    Rect2 rect2 = 5;
  }
}

message Object3 {
  Geometry3 geometry = 1;

  repeated Transform3 transforms = 3;

  reserved 2;
}

message Geometry3 {
  oneof data {
    Sphere sphere = 1;
    Mesh mesh = 2;
    Polygon3 polygon = 3;
  }
}

message Sphere {
  Number radius = 1;
}

message Mesh {
  repeated Vertex vertices = 1;

  repeated uint32 indices = 2;
  
  // TODO(doug): repeated Mesh sub_meshes = 6;
}

message Vertex {
  Point3 position = 1;
  
  Vector3 normal = 2;
}

message ArtifactUpdate {
  enum Operation {
    OPERATION_UNKNOWN = 0;
    OPERATION_CREATE = 1;
    OPERATION_UPDATE = 2;
  }

  //.google.protobuf.Any field_data = 1;

  //.google.protobuf.FieldMask field = 2;
  
  Operation operation = 3;
}

message SeriesId {
  ArtifactId artifact_id = 1;
}

// TODO(doug): Overkill? In practice, users will probably just want to use a string identifier and hard code it?
message SeriesDimensionId {
  ArtifactId artifact_id = 1;
}

message SeriesPoint {
  SeriesId series_id = 1;

  repeated SeriesDimensionValue values = 2;
}

message SeriesData {
  repeated SeriesDimensionData dimensions = 1;
}

message SeriesDimensionData {
  SeriesDimensionId id = 1;
  
  string name = 2;
}

message SeriesDimensionValue {
  SeriesDimensionId dimension_id = 1;

  Number value = 2;
}